### ISSUE TITLE
Inner and outer join with multiple join-conditions yield incorrect results

### ISSUE THREAD

**Issue by GoogleCodeExporter** [sagarswathi/h2database#203]
```

What steps will reproduce the problem?
(simple SQL scripts or simple standalone applications are preferred)


drop table base; 
drop table a; 
drop table b; 

create table a( pk integer, val varchar );
create table b( pk integer, val varchar );
create table base( pk integer, deleted integer );

insert into base values ( 1,  '0' ); 
insert into base values ( 2,  '1' ); 
insert into base values ( 3,  '0' ); 

insert into a values  ( 1, 'a' ); 
insert into b values ( 2, 'a' ); 
insert into b values ( 3, 'a' ); 

select * from a inner join base a_base on a.pk = a_base.pk left outer join
( b inner join base b_base on b.pk = b_base.pk and b_base.deleted = 0 ) 




What is the expected output? 

one row (with "deleted == 0 )


What do you see instead?

two rows (with "deleted== null and 0, respectivly)


What version of the product are you using? On what operating system, file
system, and virtual machine?

Latest (1.2.136), Windows7, no VMs, NTFS, duplicated this sample in web sql
window, orginally discovered the bug in a more complex query in a unit test


Do you know a workaround?

Yes - add the second join-condition to an explicit WHERE clause after main
query - just append WHERE b_base.deleted = 0


How important/urgent is the problem for you?

Problematic and un-fixable, but it only affects our unit tests. 

In your view, is this a defect or a feature request?

Defect. The same code yields the expected results in Oracle and Postgres. 

Please provide any additional information below.

I first assumed the LEFT OUTER JOIN is performed first, which yields
NULL-values in "deleted", and then the condition is applied. However, I
then tried adding NOT b_base.deleted IS NULL with same results. 


select * from b inner join base on b.pk = base.pk and base.deleted = 0; 

yields correct result (1 row). 

Basically, inner joins within outer joins seem rather buggy; using "1 = 2"
as boolean condition also yields 2 rows: 

select a.*, a_base.*, b.*  from a inner join base a_base on a.pk =
a_base.pk left outer join ( b inner join base b_base on 1 = 2 ) b;


```

Original issue reported on code.google.com by `MrNa...@gmail.com` on 25 May 2010 at 10:38

**Comment by GoogleCodeExporter** [comment]
```
This problem is closely related to issues 145 and 177. I have increased the 
priority
and will try to find a solution. There is a workaround, and maybe the solution 
will
be to automatically convert the query to this form:

select a.pk, a_base.pk, bpk, bbasepk  from a inner join base a_base 
on a.pk = a_base.pk left outer join
( select b.pk bpk, b_base.pk bbasepk from b inner join base b_base on b.pk =
b_base.pk and b_base.deleted = 0 ) x
on 1=1;
```

Original comment by `thomas.t...@gmail.com` on 31 May 2010 at 7:09
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
More test cases:


drop table base; 
drop table a; 
drop table b; 
create table a( pk integer, val varchar(255) );
create table b( pk integer, val varchar(255) );
create table base( pk integer, deleted integer );
insert into base values ( 1,  0 ); 
insert into base values ( 2,  1 ); 
insert into base values ( 3,  0 ); 
insert into a values  ( 1, 'a' ); 
insert into b values ( 2, 'a' ); 
insert into b values ( 3, 'a' ); 

select a.pk, a_base.pk, b.pk, b_base.pk  from a inner join base a_base 
on a.pk = a_base.pk left outer join
( b inner join base b_base on b.pk = b_base.pk and b_base.deleted = 0 ) 
on 1=1;
-- H2:
-- 1    1   2   null
-- 1    1   3   3
-- PostgreSQL, MySQL, Derby:
-- 1    1   3   3
-- HSQLDB: Unexpected token B, requires SELECT in statement

select a.pk, a_base.pk, b.pk, b_base.pk from a inner join base a_base 
on a.pk = a_base.pk left outer join
( b inner join base b_base on b.pk = b_base.pk and b_base.deleted = 0 ) ;
-- H2:
-- 1    1   2   null
-- 1    1   3   3
-- PostgreSQL: ERROR: syntax error at end of input 42601/0
-- MySQL: You have an error in your SQL syntax; 
--       check the manual that corresponds to your MySQL server version 
--       for the right syntax to use near '' at line 3
-- Derby: Syntax error: Encountered "<EOF>" at line 3, column 71. 42X01/30000
-- HSQLDB: Unexpected token B, requires SELECT in statement

select a.pk, a_base.pk, b.pk, b_base.pk  from a inner join base a_base 
on a.pk = a_base.pk left outer join
( b inner join base b_base on b.pk = b_base.pk and b_base.deleted = 0 ) 
on a.pk = b.pk;
-- H2, MySQL, PostgreSQL, Derby:
-- 1    1   null    null
-- HSQLDB: Unexpected token B, requires SELECT in statement

select a.pk, a_base.pk, bpk, bbasepk  from a inner join base a_base 
on a.pk = a_base.pk left outer join
( select b.pk bpk, b_base.pk bbasepk from b inner join base b_base on b.pk =
b_base.pk and b_base.deleted = 0 ) x
on 1=1;
-- PostgreSQL, MySQL, Derby, H2:
-- 1    1   3   3



```

Original comment by `thomas.t...@gmail.com` on 31 May 2010 at 7:12

**Comment by GoogleCodeExporter** [comment]
```
This is implemented in version 1.2.138, but not enabled by default (because 
it's a bigger change). To test it, enable the system property h2.nestedJoins. 
This property will be enabled by default in version 1.3.x
```

Original comment by `thomas.t...@gmail.com` on 28 Jun 2010 at 4:54

**Comment by GoogleCodeExporter** [comment]
```
I'm setting the issue to 'fixed', even if nested joins are not yet enabled by 
default. Please add a comment if it still doesn't work for you (or doesn't work 
as expected) even when the system property is enabled.
```

Original comment by `thomas.t...@gmail.com` on 28 Jun 2010 at 6:52
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
```
I just noticed that the one unit-test that led me to post this bug still fails 
using h2 version 1.2.140. It passes using Oracle. 

In my unit-test i explicitly do System.setProperty( "h2.nestedJoins", "true" ) 
before running the query. Do I have to set this property at start-up or should 
this be good enough? 

I will investigate further, this is a complex query that is autogenerated so it 
is a bit hard to verify this properly. 
```

Original comment by `MrNa...@gmail.com` on 3 Aug 2010 at 12:55

**Comment by GoogleCodeExporter** [comment]
```
Btw, i use prepareStatement() so it could be that the statement is cached. Will 
check this. 
```

Original comment by `MrNa...@gmail.com` on 3 Aug 2010 at 12:57

**Comment by GoogleCodeExporter** [comment]
```
Right: Setting h2.nestedJoins before any statements are cached solved the 
problem and the unit-test now passes! Thanks! 
```

Original comment by `MrNa...@gmail.com` on 3 Aug 2010 at 1:03
